### ISSUE TITLE
Warning: Operand of null-aware operation '?.' has type '<Model2>' which excludes null.

### ISSUE THREAD

**Issue by hackrx** [aws-amplify/amplify-flutter#801]
**Describe the bug**
The generated model from GraphQL schema is giving warning `Warning: Operand of null-aware operation '?.' has type 'Squad' which excludes null.` in each model.

**To Reproduce**
1. Just create two basic models in amplify admin
2. Make a 1:n or M:n relationship between two models.
3. Pull the changes
4. Then flutter run, you will see warnings in the console, and it is annoying because the terminal just gets flooded with this warning.

**Expected behavior**
I think If the value can't be null, then model gen should cover this use case
![image](https://user-images.githubusercontent.com/54453778/129147953-1df392a7-2ab3-4ad6-b0cf-3158b2a53474.png)

  In vs code
![image](https://user-images.githubusercontent.com/54453778/129148167-f3daeadd-f000-4e5d-918f-1327f4f82910.png)

We can fix this by replacing `?.` with `.`, the null check is already done on the parent, so again it should not be checked for the child.
![image](https://user-images.githubusercontent.com/54453778/129148633-b4c79b30-1882-4e4e-82bc-3d188d44ee70.png)

**Comment by HuiSF** [comment]
This fix to amplify-codegen has been released.

**Comment by hackrx** [comment]
@HuiSF Thanks for the updated fix, but I am still facing this issue. In my case, I have `1:n` and `m:n` relationships in the models, which is still showing this warning.
For eg. 
`User.dart` (Model file)
```
 Map<String, dynamic> toJson() => {
    'id': id, 'userName': _userName, 
 'userToAAA': _userToAAA?.map((e) => e?.toJson())?.toList(),   //<--------[1:n]
 'userToBBB': _userToBBB?.map((e) => e?.toJson())?.toList(),     //<-----[1:n]
 'userToCCC': _userToCCC?.map((e) => e?.toJson())?.toList(),   //<-------[1:n]
'userToDDD': _userToDDD?.map((e) => e?.toJson())?.toList(),   //<----[m:n]
 'userToEEE': _userToEEE?.map((e) => e?.toJson())?.toList(),   // <------[m:n]
  };
```
And I am facing this issue in other models too.

```
amplify_flutter: 0.2.2
```
Please let me know if the `0.2.2` version has this fix. Thanks.

**Comment by HuiSF** [comment]
Hi @hackrx Thank for the follow up!

This fix is made to amplify-codegen package, which is used by [amplify-cli](https://www.npmjs.com/package/@aws-amplify/cli) package. You may need to update the amplify-cli package installed in your local environment to get the fix and regenerate the models.

**Comment by hackrx** [comment]
@HuiSF I am already using the Upgraded Amplify CLI version
```
$amplify --version
5.3.1
$amplify upgrade                                                             ✔ 
This is the latest Amplify CLI version.
```

**Comment by HuiSF** [comment]
@hackrx it happened to me before, sometime it's pick up the latest codgen package.
Could you do this to verify?

if your `amplify-cli` is installed globally. Try run
```
npm ls -g --depth=1
```
to check which version of the amplify-codegen package was installed.

**Comment by hackrx** [comment]
Yes, it was `amplify-codegen@2.26.2`,  but now I am using the updated version.
Thanks.

**Comment by HuiSF** [comment]
Thanks for the follow up again @hackrx much appreciated! And thanks @cshfang for implementing this fix!

**Comment by hackrx** [comment]
But I am still facing this issue 😅🥲 
Here is the log of updated npm package: 
```
npm ls -g --depth=1                                                      ✔  28s 
/usr/local/lib
├─┬ @aws-amplify/cli@5.3.1
│ ├── @aws-cdk/cloudformation-diff@1.119.0
│ ├── amplify-app@3.0.8
│ ├── amplify-category-analytics@2.21.17
│ ├── amplify-category-api@2.31.19
│ ├── amplify-category-auth@2.36.3
│ ├── amplify-category-function@2.34.3
│ ├── amplify-category-hosting@2.7.17
│ ├── amplify-category-interactions@2.6.3
│ ├── amplify-category-notifications@2.19.4
│ ├── amplify-category-predictions@2.9.8
│ ├── amplify-category-storage@2.12.5
│ ├── amplify-category-xr@2.8.17
│ ├── amplify-cli-core@1.26.1
│ ├── amplify-cli-logger@1.1.0
│ ├── amplify-codegen@2.26.4
│ ├── amplify-console-hosting@1.9.8
│ ├── amplify-container-hosting@1.3.19
│ ├── amplify-dotnet-function-runtime-provider@1.6.3
│ ├── amplify-dotnet-function-template-provider@1.5.17
│ ├── amplify-frontend-android@2.15.4
│ ├── amplify-frontend-flutter@0.4.4
│ ├── amplify-frontend-ios@2.20.10
│ ├── amplify-frontend-javascript@2.23.5
│ ├── amplify-go-function-runtime-provider@1.9.0
│ ├── amplify-go-function-template-provider@1.3.10
│ ├── amplify-java-function-runtime-provider@1.8.10
│ ├── amplify-java-function-template-provider@1.5.9
│ ├── amplify-nodejs-function-runtime-provider@1.6.7
│ ├── amplify-nodejs-function-template-provider@1.6.17
│ ├── amplify-provider-awscloudformation@4.57.0
│ ├── amplify-python-function-runtime-provider@1.9.7
│ ├── amplify-python-function-template-provider@1.3.12
│ ├── amplify-util-import@1.5.8
│ ├── amplify-util-mock@3.34.1
│ ├── aws-sdk@2.980.0
│ ├── chalk@4.1.2
│ ├── ci-info@2.0.0
│ ├── cli-table3@0.6.0
│ ├── cloudform-types@4.2.0
│ ├── colors@1.4.0
│ ├── ejs@3.1.6
│ ├── enquirer@2.3.6
│ ├── env-editor@0.5.0
│ ├── esm@3.2.25
│ ├── execa@5.1.1
│ ├── folder-hash@4.0.1
│ ├── fs-extra@8.1.0
│ ├── glob@7.1.7
│ ├── global-prefix@3.0.0
│ ├── graphql-transformer-core@6.29.3
│ ├── gunzip-maybe@1.4.2
│ ├── hidefile@3.0.0
│ ├── ini@1.3.8
│ ├── inquirer@7.3.3
│ ├── lodash@4.17.21
│ ├── node-fetch@2.6.1
│ ├── open@7.4.2
│ ├── ora@4.1.1
│ ├── parse-json@5.2.0
│ ├── progress@2.0.3
│ ├── promise-sequential@1.1.1
│ ├── semver@7.3.5
│ ├── tar-fs@2.1.1
│ ├── update-notifier@4.1.3
│ ├── uuid@3.4.0
│ ├── which@2.0.2
│ └── winston@3.3.3
```
Steps which I followed: 
1. Updated the npm package globally as you suggested.
2. Deleted the old model folder from `/lib`
3. Then amplify pull --appId <id>

I checked the files, they are all same.

**Comment by HuiSF** [comment]
Hum it looks correct to me
```graphql
type Blog @model {
  id: ID!
  name: String!
  posts: [Post] @connection(keyName: "byBlog", fields: ["id"])
}
```

```dart
  Map<String, dynamic> toJson() => {
        'id': id,
        'name': _name,
        'posts': _posts?.map((Post? e) => e?.toJson()).toList()
      };
```

You can see the change of the `map` callback function, `e` is typed, and no unnecessary null aware operators.

Could you also try to execute `amplify codegen models` after pull?

**Comment by hackrx** [comment]
I tried `amplify codegen models` but still no changes in the file. [Before that, I deleted the old models folder]
![image](https://user-images.githubusercontent.com/54453778/131589416-0e436711-c431-44f6-b78f-ebb77d4587de.png)

This warning is flooding the terminal, see: 
![image](https://user-images.githubusercontent.com/54453778/131589530-7c83c11b-cbda-4071-810f-f9dcc0e032eb.png)

**Comment by HuiSF** [comment]
By any chance, @hackrx is there an instance of amplify-cli installed locally to the project? That's the only thing I can think of that's not working...

**Comment by hackrx** [comment]
@HuiSF I can see in my root project dir, I don't have any package.json there, so that means it is not installed locally to the project. anyways I did `npm uninstall @aws-amplify/cli` and then I deleted blank generated `package.json` files.
After that, I again deleted the models folder and did `amplify codegen models`, and got the same results.
To test it more I created a new flutter project, and then pulled this amplify project, and still, it is showing this warning.

**Comment by HuiSF** [comment]
Can you list me what's the scalar types are in these lists? 
e.g. `[String]`,  `[AWSDateTime]` etc.

**Comment by hackrx** [comment]
These lists are the models, as I already said I have a relationship between the models, like 1:1, 1:n, and m:n.
I am only facing this issue with 1:n and m:n relatioship.
For example, I have a `Like` model like this: 
![image](https://user-images.githubusercontent.com/54453778/131954623-40777457-a224-49f3-969a-bec17585773b.png)

and In the `User` model, I have a relationship with `Like` model as `1:n` i.e one user can have many likes.
![image](https://user-images.githubusercontent.com/54453778/131954672-da03d095-3a2b-4a8b-ad95-bf75cdb99a7a.png)

In the `User.dart` file, it is showing this:
![image](https://user-images.githubusercontent.com/54453778/131954554-d138227f-0d5c-4189-98a8-d05e139eace9.png)

**Comment by HuiSF** [comment]
Got it, hum that's weird, I couldn't reproduce this on my end 😅

**Comment by hackrx** [comment]
Can you please reopen this issue? because I am still facing this issue. and I am sure this issue still exists.
did you tried making a model which has a relationship of 1:n, and m:n?
like this one:
![image](https://user-images.githubusercontent.com/54453778/132062544-47ef08f7-b900-43ff-a465-438b66302955.png)
Please try to create the same model, and If it is working for you, please let me know.
Thanks.

**Comment by hackrx** [comment]
@HuiSF can you please confirm if this model [Above sample model] is working fine for you?

**Comment by HuiSF** [comment]
@hackrx 

Here's the `toJson` method in the generated model of `QuoteUser` which I believe is what shown in your screenshots.

```
  Map<String, dynamic> toJson() => {
        'id': id,
        'name': _name,
        'email': _email,
        'phoneNumber': _phoneNumber,
        'ImageURL': _ImageURL,
        'QuoteUserToPosts':
            _QuoteUserToPosts?.map((QuotePost? e) => e?.toJson()).toList(),
        'QuoteUserToLikes':
            _QuoteUserToLikes?.map((QuoteLike? e) => e?.toJson()).toList(),
        'QuoteUserToComments':
            _QuoteUserToComments?.map((QuoteComment? e) => e?.toJson()).toList()
      };
```

**Comment by hackrx** [comment]
@HuiSF thanks for looking into this issue. So now it is reproducible at your end too. Please reopen this issue, because it is not fixed yet. 
Thanks.

**Comment by HuiSF** [comment]
No @hackrx 😅 please take a look closely. The generated I pasted is different from your screenshot. 

```
(QuotePost? e) => e?.toJson()
```

Note that the parameter `e` is typed as nullable `QuotePost` which removes the warning raised from `e?.toJson()`. So I suspect that somewhere in your development environment may have multiple version of amplify cli.

**Comment by hackrx** [comment]
Thanks, @HuiSF for the follow-up. I solved this issue by following these steps: 
 First, since there was an old version of amplify-cli installed on my system, I uninstalled it globally using 
`npm uninstall -g @aws-amplify/cli`, but still in my console `amplify --version ` was showing `5.3.1` then I did 
`amplify uninstall`, after there was no output for `-version`. 
Then I did `npm install -g @aws-amplify/cli`
`amplify -v` gave `5.5.0`. And after that, I again pulled the test project, and it fixed this warning.
